rego/rego_test.go

```diff
@@ -492,6 +492,59 @@ func TestPrepareAndEvalNewTransaction(t *testing.T) {
 	assertPreparedEvalQueryEval(t, pq, []EvalOption{EvalTransaction(newTxn)}, "[[2]]")
 }
 
+func TestPrepareAndEvalNoTransactionOpt(t *testing.T) {
+	module := `
+	package test
+	x = data.foo.y
+	`
+	ctx := context.Background()
+	store := inmem.New()
+	txn := storage.NewTransactionOrDie(ctx, store, storage.WriteParams)
+
+	path, ok := storage.ParsePath("/foo")
+	if !ok {
+		t.Fatalf("Unexpected error parsing path")
+	}
+
+	err := storage.MakeDir(ctx, store, txn, path)
+	if err != nil {
+		t.Fatalf("Unexpected error writing to store: %s", err.Error())
+	}
+
+	err = store.Write(ctx, txn, storage.AddOp, path, map[string]interface{}{"y": 1})
+	if err != nil {
+		t.Fatalf("Unexpected error writing to store: %s", err.Error())
+	}
+
+	r := New(
+		Query("data.test.x"),
+		Module("", module),
+		Store(store),
+		Transaction(txn),
+	)
+
+	pq, err := r.PrepareForEval(ctx)
+	if err != nil {
+		t.Fatalf("Unexpected error: %s", err.Error())
+	}
+
+	assertPreparedEvalQueryEval(t, pq, []EvalOption{EvalTransaction(txn)}, "[[1]]")
+	store.Commit(ctx, txn)
+
+	// Update the store
+	newTxn := storage.NewTransactionOrDie(ctx, store, storage.WriteParams)
+	err = store.Write(ctx, newTxn, storage.ReplaceOp, path, map[string]interface{}{"y": 2})
+	if err != nil {
+		t.Fatalf("Unexpected error writing to store: %s", err.Error())
+	}
+	store.Commit(ctx, newTxn)
+
+	// Expect that the old transaction and new (auto-created) transaction
+	// give different results
+	assertPreparedEvalQueryEval(t, pq, []EvalOption{EvalTransaction(txn)}, "[[1]]")
```

Comment:
This surprised me. We don't have a bit on the transaction to detect read/write operations on committed/aborted transactions which means that reads performed on these committed/aborted transactions are not isolated. We probably shouldn't allow them.

Not sure what the conclusion should be here. We could change the test to mock out the store and then assert on the # of transactions that get opened but that's as valuable.

Reply:
Yea.. i was wondering about this behavior. I added the test primarily so we can know if it changes. I like the idea of a mock store though, asserting the number of transactions and, in theory, reads/writes on it seems pretty useful.

For this test what we really want to check is whether or not it is a) using the txn passed in, and b) auto creating a txn and using it all the way through eval. For those purposes I think checking the # of created txn's is probably fine. For (a) we can make sure it is 0 and (b) we can make sure it is 1. Anything different is a problem.